miden-crypto/src/word.rs

```diff
@@ -0,0 +1,626 @@
+    }
+}
+
+impl Hash for Word {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write(&self.as_bytes());
+    }
+}
+
+impl Digest for Word {
+    fn as_bytes(&self) -> [u8; WORD_SIZE_BYTES] {
+        self.as_bytes()
+    }
+}
+
+impl Deref for Word {
+    type Target = [Felt; WORD_SIZE_FELT];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Ord for Word {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Compare the inner u64 of both elements.
+        //
+        // It will iterate the elements and will return the first computation different than
+        // `Equal`. Otherwise, the ordering is equal.
+        //
+        // Finally, we use `Felt::inner` instead of `Felt::as_int` so we avoid performing a
+        // montgomery reduction for every limb. That is safe because every inner element of the
+        // word is guaranteed to be in its canonical form (that is, `x in [0,p)`).
+        //
+        // Because we don't perform Montgomery reduction, we must iterate over, and compare,
+        // each element. A simple bytestring comparison would be inappropriate because the `Word`s
+        // are represented in "lexicographical" order.
+        self.0.iter().map(Felt::inner).zip(other.0.iter().map(Felt::inner)).fold(
+            Ordering::Equal,
+            |ord, (a, b)| match ord {
+                Ordering::Equal => a.cmp(&b),
+                _ => ord,
+            },
+        )
+    }
+}
+
+impl PartialOrd for Word {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Display for Word {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{}", self.to_hex())
+    }
+}
+
+impl Randomizable for Word {
+    const VALUE_SIZE: usize = WORD_SIZE_BYTES;
+
+    fn from_random_bytes(bytes: &[u8]) -> Option<Self> {
+        let bytes_array: Option<[u8; 32]> = bytes.try_into().ok();
+        if let Some(bytes_array) = bytes_array {
+            Self::try_from(bytes_array).ok()
+        } else {
+            None
+        }
+    }
+}
+
+// CONVERSIONS: FROM WORD
+// ================================================================================================
+
+/// Errors that can occur when working with a [Word].
+#[derive(Debug, Error)]
+pub enum WordError {
+    /// Failed to convert the word's field elements to the specified type.
+    #[error("failed to convert the word's field elements to type {0}")]
+    TypeConversion(&'static str),
+    /// Field element conversion failed due to invalid value.
+    #[error("failed to convert to field element: {0}")]
+    InvalidFieldElement(String),
+    /// Field elements parsed are out of range.
+    #[error("values of a Word must be inside the field modulus")]
+    OutOfRange,
+    /// Hex-encoded field elements parsed are invalid.
+    #[error("hex encoded values of a word are invalid")]
+    HexParse(#[from] HexParseError),
+}
```

Comment:
@bobbinth I've added OutOfRange and HexParse here when replacing the direct `type Error = HexParseError;` usage.